config/initializers/proxy_uri.rb

```diff
@@ -0,0 +1,41 @@
+#
+# The Ruby URI parser doesn't decode the percent encoded characters in the URI, in particular it
+# doesn't decode the password which is frequently used when specifying proxy addresses and
+# authentication. For example, the following code:
+#
+#   require 'uri'
+#   proxy = URI.parse('http://myuser:%24%3fxxxx@192.168.100.10:3128')
+#   puts proxy.password
+#
+# Produces the following output:
+#
+#   %24%3fxxxx
+#
+# But some gems, in particular `rest-client` and `kubeclient`, expect it to decode those characters,
+# as they use the value returned by the `password` method directly, and thus they fail to
+# authenticate against the proxy server when the password contains percent encoded characters.
+#
+# To address this issue this file adds a new `proxy` URI schema that almost identical to the `http`
+# schema, but that decodes the password before returning it. Users can use this schema instead of
+# `http` when they need to use percent encoded characters in the password. For example, the user
+# can type in the UI the following proxy URL:
+#
+#   proxy://myuser:%24%3fxxxx@192.168.100.10:3128
+#
+# And the new schema will automatically translate `%24%3fxxxx` into `$?xxxx`.
+#
+
+require 'cgi'
+require 'uri'
+
+module URI
+  class PROXY < HTTP
+    def password
```

Comment:
we might also want to handle percent-encoded username
(RFC just says whole userinfo field may contain `pct-encoded` chars; username encoding does happen in the wild for example when using email as username https://stackoverflow.com/questions/10050877/url-username-with)

Reply:
Yes, that makes sense. Can you comment that also in the `rest-client` and `kubeclient` issues?

Reply:
Are you monkey patching an existing class?  If so I am against this approach.  This is a known thing about the URI class, and it's the callers responsibility to do the CGI escape/unescape thing.

I would not mind if you created a subclass, though not sure what it would be named.  Id  also argue that new subclass should live in a utility library like more_core_extensions or manageiq_gems_pending.

@bdunne @jrafanie  Thoughts?

Reply:
I'm not a fan of the name `PROXY`, but I like the idea.
> Id also argue that new subclass should live in a utility library like `more_core_extensions`

:+1:

Reply:
I agree, `PROXY` is confusing to me.

There's already `find_proxy` and `use_proxy` methods that leverage environment variables HTTP_PROXY, FTP_PROXY, NO_PROXY, which are completely different things than this.  See [here](https://github.com/ruby/ruby/blob/506512c028b0967ee04239d17a92ccb42b3241b1/lib/uri/generic.rb#L1489-L1566).